python/tank/descriptor/bundle_cache_usage/logger.py

```diff
@@ -0,0 +1,281 @@
+        # TODO: NICOLAS: returning would cause a silent non-usage of specified parameter
+        if (self.__initialized):
+            return
+        self.__initialized = True
+        tmp_log.debug_worker_threading("__init__")
+        self._terminate_requested = threading.Event()
+        self._queued_signal = threading.Event()
+        self._tasks = Queue.Queue()
+        self._errors = Queue.Queue()
+        self._member_lock = threading.Condition()
+        self._completed_count = 0
+        self._pending_count = 0
+        # Make use of database class path validation.
+        # This way we can catch and log such an error up-front
+        # before the worker thread is actually started.
+        BundleCacheUsageDatabase(bundle_cache_root)
+        self._bundle_cache_root = bundle_cache_root
+
+    @ classmethod
+    def delete_instance(cls, timeout=DEFAULT_STOP_TIMEOUT):
+        """
+        Safely lock the instance, stop the worker thread and set the instance to None
+        :param timeout: A float timeout in seconds
+        """
+        with cls.__singleton_lock:
+            if cls.__singleton_instance:
+                cls.__singleton_instance.stop(timeout)
+                cls.__singleton_instance = None
+
+    #
+    # Private methods
+    # Running exclusively from the worker thread context
+    #
+    ###########################################################################
+
+    def __log_error(self, exception, message):
+        """
+        Helper method that adds an exception and custom message to a queue allowing
+        reporting of errors from the worker thread back into the main thread.
+
+        :param exception: A
+        :param message: a str custom message
+        """
+        self._errors.put((exception, message))
+        self._errors.task_done()
+
+    def __log_usage(self, bundle_path):
+        """
+        Update database's last usage date and usage count for the specified entry.
+
+        .. note:: Invoked exclusively from the worker thread.
+
+        :param bundle_path: A str of a bundle path
+        """
+        database = BundleCacheUsageDatabase(self._bundle_cache_root)
+        database.log_usage(bundle_path)
+
+    def __consume_task(self):
+        """
+        Execute a queued task.
+
+        .. note:: Invoked exclusively from the worker thread.
+        """
+        try:
+            method, args, kwargs = self._tasks.get()
+            if method:
+                method(*args, **kwargs)
+        except Exception as e:
+            self.__log_error(e, "Unexpected error consuming task")
+
+        with self._member_lock:
+            if self._pending_count > 0:
+                self._pending_count -= 1
+            self._completed_count += 1
+
+    def run(self):
+        """
+        Implementation of threading.Thread.run()
+        """
+
+        tmp_log.debug_worker_threading("starting")
+
+        try:
+            while not self._terminate_requested.is_set() or self.pending_count > 0:
```

Comment:
do we need _pending count? could we make it simpler and just let the queue handle the locking and the pop a "QUIT" message on the queue if we want to terminate? That way we could eliminate a bunch of the lock logic and the shared variables?

could it be as simple as 

```
try:
   while True:
      # blocking
      method, args, kwargs = self._tasks.get()
      if method == "QUIT":
          break
except ...
```